pkg/p2p/kadcast/protocol.go

```diff
@@ -0,0 +1,73 @@
+package kadcast
+
+import (
+	"errors"
+	"log"
+	"time"
+)
+
+// InitBootstrap inits the Bootstrapping process by sending 
+// a `PING` message to every bootstrapping node repeatedly.
+// If it tried 3 or more times and no new `Peers` were added,
+// it panics. 
+// Otherways, it returns `nil` and logs the Number of peers 
+// the node is connected to at the end of the process.
+func InitBootstrap(router *Router, bootNodes []Peer) error {
+	log.Println("Bootstrapping process started.")
+	// Get PeerList ordered by distance so we can compare it
+	// after the `PONG` arrivals.
+	initPeerNum := router.tree.getTotalPeers()
+	for i := 0; i < 3; i++ {
+		// Send `PING` to the bootstrap nodes.
+		for _, peer := range bootNodes {
+			router.sendPing(peer)
+		}
+		// Wait for `PONG` responses.
+		time.Sleep(time.Second * 5)
+		// If new peers were added (the bootstrap nodes)
+		// we consider that the bootstrapping succeeded.
+		actualPeers := router.tree.getTotalPeers()
+		if actualPeers <= initPeerNum {
+			if i < 3 {
+				log.Printf("Bootstrapping nodes were not added on attempt nº %v\nTrying again...\n", i)
+			} else {
+				return errors.New("\nMaximum number of attempts achieved. Please review yor connection settings\n")
+			}
+		} else {
+			break
+		}
+	}
+	log.Printf("Bootstrapping process finnished. \nYou're now connected to %v nodes", router.tree.getTotalPeers())
+	return nil
+}
+
+// StartNetworkDiscovery triggers the network discovery process.
+// The node basically sends `FIND_NODES` messages to the nodes it
+// is currently connected to and evaluates the `Peers` that were added 
+// on each iteration. 
+// If the closest peer to ours is the same during two iterations of the
+// `FIND_NODES` message, we finnish the process logging the ammout of peers
+// we are currently connected to.
+// Otherways, if the closest Peer on two consecutive iterations changes, we 
+// keep queriyng the `alpha` closest nodes with `FIND_NODES` messages.
+func StartNetworkDiscovery(router *Router) {
+	var actualClosest []Peer
+	previousClosest := router.getXClosestPeersTo(1, router.MyPeerInfo)
+	// Ask for nodes to `alpha` closest nodes to my peer.
+	router.sendFindNodes()
+	// Wait until response arrives and we query the nodes.
+	time.Sleep(time.Second * 5)
```

Comment:
I just tried to keep the protocol as lightweight as possible. 

And note that here we don't have any way to identify if some packet is a response of another one since packets are not related to any ID or basically anything.

At this point it's complicated to do it like this since I will never know when a packet is a response to a previous one I sent. Also, it's important to consider that it will add complexity and force us to work with more signals and conditions just for make it "interactive" with the packet arrivals.

I agree it does not look cool, but it's effective and lightweight. 
We can discuss the implementations that we could do.

Reply:
I am not convinced by your explanation. Either the process is synchronous (I send a `findNodes` request and I wait for some kind of side effect, like a response that may or may not be related to the message I broadcasted) and therefore the process should block until some condition is met (whatever the condition). In case the process is instead asynchronous, `sendFindNodes` should follow a fire-and-forget model (and therefore there should be no followup in the same thread/goroutine basically). In both cases, the `sleep` is inappropriate.